src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -44,52 +44,51 @@ private[concurrent] trait BatchingExecutor extends Executor {
   // invariant: if "_tasksLocal.get ne null" then we are inside BatchingRunnable.run; if it is null, we are outside
   private val _tasksLocal = new ThreadLocal[List[Runnable]]()
 
-  private class Batch(val initial: List[Runnable]) extends Runnable with BlockContext {
+  private class Batch(val initial: List[Runnable]) extends Runnable with BlockContext with (BlockContext => Unit) {
     private var parentBlockContext: BlockContext = _
     // this method runs in the delegate ExecutionContext's thread
-    override def run(): Unit = {
-      require(_tasksLocal.get eq null)
-
-      val prevBlockContext = BlockContext.current
-      BlockContext.withBlockContext(this) {
-        try {
-          parentBlockContext = prevBlockContext
+    override def run(): Unit = BlockContext.usingBlockContext(this)(BlockContext.current)(this)
 
-          @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
-            case Nil => ()
-            case head :: tail =>
-              _tasksLocal set tail
-              try {
-                head.run()
-              } catch {
-                case t: Throwable =>
-                  // if one task throws, move the
-                  // remaining tasks to another thread
-                  // so we can throw the exception
-                  // up to the invoking executor
-                  val remaining = _tasksLocal.get
-                  _tasksLocal set Nil
-                  unbatchedExecute(new Batch(remaining)) //TODO what if this submission fails?
-                  throw t // rethrow
-              }
-              processBatch(_tasksLocal.get) // since head.run() can add entries, always do _tasksLocal.get here
-          }
+    override def apply(prevBlockContext: BlockContext): Unit = {
+      require(_tasksLocal.get eq null)
+      try {
+        parentBlockContext = prevBlockContext
 
-          processBatch(initial)
-        } finally {
-          _tasksLocal.remove()
-          parentBlockContext = null
+        @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
+          case Nil => ()
+          case head :: tail =>
```

Comment:
If you want to be really efficient, you can optimise the common 1-task case by storing a `Runnable` directly instead of a single element, then fall back to a `List[Runnable]` when you need more than one element. Something like [this code from Play](https://github.com/playframework/playframework/blob/58a150d712bdf3bf4d14709569afcfd9ea97fb7f/framework/src/iteratees/src/main/scala/play/api/libs/iteratee/Execution.scala#L33-L125).